Reintroduce iAlarm integration

RyuzakiKK opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Previous iAlarm integration has been removed because it used
webscraping https://github.com/home-assistant/core/pull/43010.

Now pyialarm library has been updated to the version 1.0 that uses the iAlarm API instead https://github.com/RyuzakiKK/pyialarm/blob/master/pyialarm/pyialarm.py

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

This is configurable only with the new config flow.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15718

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RyuzakiKK commented:
Added the documentation PR https://github.com/home-assistant/home-assistant.io/pull/15718

bdraco reviewed (changes requested):
Per ADR0010, https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md new integrations should be a config flow.


`python3 -m script.scaffold config_flow` can help you get started.

RyuzakiKK commented:
Thanks @bdraco for the review. I'm on it.

RyuzakiKK commented:
@bdraco I rewrote the iAlarm integration and now it uses the new config flow.

I flagged it as a "silver" score, but please double check if that's correct or if I missed something.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+    def __init__(self, coordinator, code):
+        """Init the iAlarm device."""
+        super().__init__(coordinator)
+        self._unique_name = f"ialarm_{coordinator.host}"
```
This unique id does not meet the unique id requirements

https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

If the device doesn't have a serial number or mac you can access use `config_entry.entry_id` instead of the host

RyuzakiKK replied:
I changed the ialarm library to expose the device mac too. Now I use it as the unique name.

bdraco commented on homeassistant/components/ialarm/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+        try:
+            result = await _validate_connection(self.hass, host, port)
+            if result is None:
+                errors["base"] = "cannot_connect"
```
Please add coverage for this line

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+
+    def alarm_disarm(self, code=None):
+        """Send disarm command."""
+        if not self._validate_code(code):
+            _LOGGER.warning("Wrong code entered for disarming")
+            return
```
```suggestion
        if not self._validate_code(code):
            _LOGGER.warning("Wrong code entered for disarming")
            return
```

It doesn't look like the device requires a code.

RyuzakiKK replied:
@bdraco you are right, the device doesn't require a code, but it is used in HA to prevent disarming with a single button click.
Do you think I should still remove it?

bdraco replied:
We shouldn't fake support for codes in Home Assistant since the underly device doesn't actually require a code.  We don't want to give the impression that the code is a security control since there is no actual validation by the device.

RyuzakiKK replied:
I see your point, it sounds good to me.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+
+    def alarm_arm_home(self, code=None):
+        """Send arm home command."""
+        if not self._validate_code(code):
+            _LOGGER.warning("Wrong code entered for arming")
+            return
```
```suggestion
        if not self._validate_code(code):
            _LOGGER.warning("Wrong code entered for arming")
            return
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+
+    def alarm_arm_away(self, code=None):
+        """Send arm away command."""
+        if not self._validate_code(code):
+            _LOGGER.warning("Wrong code entered for arming")
+            return
```
```suggestion
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+    def supported_features(self) -> int:
+        """Return the list of supported features."""
+        return SUPPORT_ALARM_ARM_HOME | SUPPORT_ALARM_ARM_AWAY
+
+    @property
+    def code_format(self):
+        """Return one or more digits/characters."""
+        return FORMAT_NUMBER
+
+    def _validate_code(self, code):
+        """Validate given code."""
+        if self._code is None:
+            return True
+        return code == self._code
+
```
```suggestion
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+async def async_setup_entry(hass, entry, async_add_entities) -> None:
+    """Set up a iAlarm alarm control panel based on a config entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id][DATA_COORDINATOR]
+    pin = entry.data.get(CONF_PIN)
+    async_add_entities([IAlarmPanel(coordinator, pin)], False)
```
```suggestion
    async_add_entities([IAlarmPanel(coordinator)], False)
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+class IAlarmPanel(AlarmControlPanelEntity, CoordinatorEntity):
+    """Representation of an iAlarm device."""
+
+    def __init__(self, coordinator, code):
```
```suggestion
    def __init__(self, coordinator):
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,95 @@
+        """Init the iAlarm device."""
+        super().__init__(coordinator)
+        self._unique_name = f"ialarm_{coordinator.host}"
+        self._code = code
```
```suggestion
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,114 @@
+        else:
+            state = None
+
+        self.state = state
+
+        return {
+            "state": state,
+        }
```
Please store the state only in one place, we shouldn't have two sources of truth

RyuzakiKK replied:
Right, now I store the state only as a variable.

bdraco commented on homeassistant/components/ialarm/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+        vol.Optional(CONF_PIN): str,
```
```suggestion
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+import voluptuous as vol
+
+from homeassistant import config_entries, core
+from homeassistant.const import CONF_HOST, CONF_PIN, CONF_PORT
```
```suggestion
from homeassistant.const import CONF_HOST, CONF_PORT
```

It doesn't look like the device requires a code.

bdraco commented on homeassistant/components/ialarm/const.py:
```diff
@@ -0,0 +1,8 @@
+DATA_COORDINATOR = "ialarm"
+
+DEFAULT_PORT = 18034
+DEFAULT_SCAN_INTERVAL = 30
```
```suggestion
```

unused

bdraco commented on homeassistant/components/ialarm/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "codeowners": [
+    "@RyuzakiKK"
+  ],
+  "config_flow": true
```
This is probably discoverable via dhcp

https://developers.home-assistant.io/docs/creating_integration_manifest/#dhcp

RyuzakiKK replied:
I added the dhcp discovery. Unfortunately the device doesn't provide an hostname, so I had to use only the MAC address.

Probably that's fine, but if this is too broad I can remove the dhcp discovery.

bdraco replied:
`async_step_dhcp` needs to be implemented for this to pass the ip to the integration.

Its ok to remove it for now and add it in a future PR

RyuzakiKK replied:
Okay, I removed it so that the merge of this integration can be faster. And after it will be included in HA I can create a separate PR to add the dhcp discoverability.

bdraco reviewed (changes requested):
Please see comments above

RyuzakiKK commented:
Sorry for the bump, but it would be great if someone could have a look at this PR to approve it or suggest the necessary changes.
Thanks!

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    if mac is None:
+        raise ConfigEntryNotReady
+
+    coordinator = IAlarmDataUpdateCoordinator(
+        hass,
+        ialarm,
+        mac,
+    )
+
+    await coordinator.async_refresh()
+
+    if coordinator.state is None:
+        raise ConfigEntryNotReady
```
```suggestion
    coordinator = IAlarmDataUpdateCoordinator(hass, ialarm, mac)
    await coordinator.async_config_entry_first_refresh()

```

RyuzakiKK replied:
Ah this is brand new!

bdraco replied:
It looks like this was not addressed.

RyuzakiKK replied:
Ops, I guess I amended my last commit without staging the changes. Sorry about that.

It should be okay now.

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,124 @@
+
+    if coordinator.state is None:
+        raise ConfigEntryNotReady
+
```
```suggestion
    hass.data.setdefault(DOMAIN, {})

```

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,124 @@
+PLATFORM = "alarm_control_panel"
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up iAlarm components."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
```
```suggestion
```

Not needed after rebase

bdraco replied:
It looks like this was not addressed.

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,69 @@
+            "manufacturer": "Antifurto365 - Meian",
+        }
+
+    @property
+    def unique_id(self):
+        """Return a unique id."""
+        return self._unique_name
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self.unique_id
```
Are these reversed?

RyuzakiKK replied:
I guess so, should be better now.

bdraco reviewed (changes requested):
Please see comments above

RyuzakiKK commented:
@bdraco Thanks for the new review, I addressed the things you pointed out.

bdraco reviewed (changes requested):
Please address outstanding comments. Thanks

RyuzakiKK commented:
@bdraco Now I staged the changes that I forgot to add.

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,112 @@
+async def _get_device_mac(hass: HomeAssistant, ialarm) -> str:
+    try:
+        async with timeout(10):
+            return await hass.async_add_executor_job(ialarm.get_mac)
```
Verified this it obtained via the api 👍

bdraco replied:
We shouldn't raise `UpdateFailed` here as thats only for the coordinator.

RyuzakiKK replied:
Ah okay, fixed it with the latest revision.

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
+    ialarm = IAlarm(host, port)
+    mac = await _get_device_mac(hass, ialarm)
+
+    if mac is None:
+        raise ConfigEntryNotReady
```
```suggestion
    try:
        async with timeout(10):
            return await hass.async_add_executor_job(ialarm.get_mac)
	except(asyncio.TimeoutError, ConnectionError) as ex
        raise ConfigEntryNotReady from ex
```

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,112 @@
+PLATFORM = "alarm_control_panel"
+_LOGGER = logging.getLogger(__name__)
+
+
+async def _get_device_mac(hass: HomeAssistant, ialarm) -> str:
+    try:
+        async with timeout(10):
+            return await hass.async_add_executor_job(ialarm.get_mac)
+    except ConnectionError as error:
+        raise UpdateFailed(error) from error
+
```
```suggestion
```

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,69 @@
+    async_add_entities([IAlarmPanel(coordinator)], False)
+
+
+class IAlarmPanel(AlarmControlPanelEntity, CoordinatorEntity):
```
```suggestion
class IAlarmPanel(CoordinatorEntity, AlarmControlPanelEntity):
```

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,69 @@
+    def __init__(self, coordinator):
+        """Init the iAlarm device."""
+        super().__init__(coordinator)
+        self._unique_id = f"ialarm_{coordinator.mac}"
```
No need to add a prefix

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,69 @@
+    @property
+    def unique_id(self):
+        """Return a unique id."""
+        return self._unique_id
```
```suggestion
        return self.coordinator.mac
```

bdraco commented on homeassistant/components/ialarm/alarm_control_panel.py:
```diff
@@ -0,0 +1,69 @@
+    @property
+    def name(self):
+        """Return the name."""
+        return self.unique_id
```
```suggestion
        return "iAlarm"
```

bdraco commented on homeassistant/components/ialarm/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+
+async def _get_device_mac(hass: core.HomeAssistant, host, port):
+    ialarm = IAlarm(host, port)
+    mac = await hass.async_add_executor_job(ialarm.get_mac)
+    return mac
```
```suggestion
    return await hass.async_add_executor_job(ialarm.get_mac)
```

bdraco commented on homeassistant/components/ialarm/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+            # If we are able to get the MAC address, we are able to establish
+            # a connection to the device.
+            mac = await _get_device_mac(self.hass, host, port)
+            if mac is None:
+                errors["base"] = "cannot_connect"
```
Can this happen without it throwing?

RyuzakiKK replied:
I'm not 100% sure about that. I guess that there is the remote possibility that the alarm sends a malformed but valid reply that doesn't contain the MAC address that we asked. In that case `get_mac()` will return `None` without throwing.

I don't know if this is something that might happen in the real world, but probably it's better be safe and check for `None`?

bdraco replied:
Ideally if there is malformed data coming back from the device there is an exception thrown by the underlying library

RyuzakiKK replied:
Right, I tagged a new release of the pyialarm library. Now if it doesn't receive the expected parameter from the alarm, a ConnectionError will be raised.
For this reason I removed the `if mac is None` check here.

bdraco commented on homeassistant/components/ialarm/strings.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "title": "Antifurto365 iAlarm",
```
```suggestion
```

not needed since its the same as manifest.json

bdraco reviewed (changes requested):
Please see new comments above

RyuzakiKK commented:
96.10% of code coverage target? I guess I should add some more tests :)

bdraco reviewed (changes requested):
Please write the missing tests (https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations) or add files not covered by tests to `.coveragerc` (config_flow.py is required to 100% coverage before merging)

bdraco commented on homeassistant/components/ialarm/__init__.py:
```diff
@@ -0,0 +1,107 @@
+        status = self.ialarm.get_status()
+        _LOGGER.debug("iAlarm status: %s", status)
+
+        if status == self.ialarm.DISARMED:
+            state = STATE_ALARM_DISARMED
+        elif status == self.ialarm.ARMED_AWAY:
+            state = STATE_ALARM_ARMED_AWAY
+        elif status == self.ialarm.ARMED_STAY:
+            state = STATE_ALARM_ARMED_HOME
+        elif status == self.ialarm.TRIGGERED:
+            state = STATE_ALARM_TRIGGERED
+        else:
+            state = None
```
```suggestion
        if status == self.ialarm.DISARMED:
            state = STATE_ALARM_DISARMED
        elif status == self.ialarm.ARMED_AWAY:
            state = STATE_ALARM_ARMED_AWAY
        elif status == self.ialarm.ARMED_STAY:
            state = STATE_ALARM_ARMED_HOME
        elif status == self.ialarm.TRIGGERED:
            state = STATE_ALARM_TRIGGERED
        else:
            state = None
```

Make a dictionary that maps ialarm state to hass state.

IALARM_TO_HASS = {
    self.ialarm.DISARMED: STATE_ALARM_DISARMED,
    ...
}

Then you can do

state = IALARM_TO_HASS.get(status)

RyuzakiKK replied:
Sure, added in the latest revision.

bdraco commented on tests/components/ialarm/test_init.py:
```diff
@@ -0,0 +1,65 @@
+
+    assert entry_id in hass.data[DOMAIN]
+    await async_unload_entry(hass, config_entry_mock)
+    assert entry_id not in hass.data[DOMAIN]
```
Don't call `async_unload_entry` or `async_setup_entry` use the standard interfaces in  https://github.com/home-assistant/core/blob/4cce724473233d4fb32c08bd251940b1ce2ba570/homeassistant/config_entries.py#L693

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

RyuzakiKK replied:
I checked a couple of recently accepted PR and I changed this test accordingly. Now hopefully it should be okay.

RyuzakiKK committed:
Introduce support for iAlarm systems

The previous iAlarm integration has been removed because it used
webscraping #43010.

Since then, the pyialarm library has been updated to use the iAlarm API
instead.

With this commit I reintroduce the iAlarm integration, leveraging the
new HA config flow.

Signed-off-by: Ludovico de Nittis <ludovico.denittis@collabora.com>

bdraco reviewed (approved):
Lgtm.  Thanks!

MartinHjelmare commented on homeassistant/components/ialarm/strings.json:
```diff
@@ -0,0 +1,20 @@
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "port": "[%key:common::config_flow::data::port%]",
+          "pin": "[%key:common::config_flow::data::pin%]"
```
The `pin` string isn't in the user form schema.

RyuzakiKK replied:
By the end of today I'll open a PR to fix them.

MartinHjelmare commented on tests/components/ialarm/test_init.py:
```diff
@@ -0,0 +1,85 @@
+    await async_setup_component(
+        hass,
+        DOMAIN,
+        {
```
We don't support config yaml for this integration. Passing the config here doesn't do anything. Set up the integration with `hass.config_entries.async_setup` instead.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
